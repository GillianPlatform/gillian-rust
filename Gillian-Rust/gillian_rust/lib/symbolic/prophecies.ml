open Monadic
module DR = Delayed_result
open DR.Syntax
open Delayed.Syntax
open Heap
open Gil_syntax
open Delayed_utils

type prophecy = {
  value : Expr.t;
  obs_ctl : Expr.t;
  observer : bool;
  controller : bool;
}

let prophecy_lvars prophecy =
  let open Utils.Containers in
  SS.union (Expr.lvars prophecy.obs_ctl) (Expr.lvars prophecy.value)

let pp_prophecy =
  let open Fmt in
  Fmt.braces
  @@ record ~sep:semi
       [
         field "value" (fun x -> x.value) Expr.pp;
         field "obs_ctl" (fun x -> x.obs_ctl) Expr.pp;
         field "observer" (fun x -> x.observer) Fmt.bool;
         field "controller" (fun x -> x.controller) Fmt.bool;
       ]

type t = prophecy MemMap.t

let sure_is_nonempty =
  MemMap.exists (fun _ { observer; controller; _ } -> observer || controller)

let assertions (pcies : t) =
  let cps loc pcy =
    (* Location and ty are uniform *)
    let loc = Expr.loc_from_loc_name loc in
    (* The future value of the prophecy variable *)
    let value =
      let cp = Actions.cp_to_name Pcy_value in
      Asrt.GA (cp, [ loc ], [ pcy.value ])
    in
    let controller =
      match pcy.controller with
      | false -> []
      | true ->
          let cp = Actions.cp_to_name Pcy_controller in
          [ Asrt.GA (cp, [ loc ], [ pcy.obs_ctl ]) ]
    in
    let observer =
      match pcy.observer with
      | false -> []
      | true ->
          let cp = Actions.cp_to_name Value_observer in
          [ Asrt.GA (cp, [ loc ], [ pcy.obs_ctl ]) ]
    in
    value :: (controller @ observer)
  in
  MemMap.fold (fun loc pcy acc -> cps loc pcy @ acc) pcies []

let lvars t =
  let open Utils.Containers in
  MemMap.fold (fun _ pcy acc -> SS.union acc (prophecy_lvars pcy)) t SS.empty

let empty = MemMap.empty
let to_yojson _ = `Null
let of_yojson _ = Error "Heap.of_yojson: Not implemented"

let cons_value_obs pcy_env pcy_var =
  match MemMap.find_opt pcy_var pcy_env with
  | Some ({ observer = true; obs_ctl; _ } as pcy) ->
      (* We consume the pcy *)
      let new_pcy_env =
        MemMap.add pcy_var { pcy with observer = false } pcy_env
      in
      DR.ok (obs_ctl, new_pcy_env)
  | Some { observer = false; _ } -> DR.error (Err.Missing_observer pcy_var)
  | None -> DR.error (Err.Missing_pcy pcy_var)

let prod_value_obs pcy_env pcy_id obs_value =
  let+ new_block =
    match MemMap.find_opt pcy_id pcy_env with
    | Some { observer = true; _ } -> Delayed.vanish () (* Duplicated resource *)
    | Some { value; obs_ctl; observer = false; controller = true } ->
        (* We already have a controller, we learn equality *)
        let learned = [ Formula.Infix.( #== ) obs_value obs_ctl ] in
        Delayed.return ~learned
          { value; observer = true; controller = true; obs_ctl }
    | Some { value; observer = false; controller = false; obs_ctl = _ } ->
        (* We have neither the controller nor the observer, the obs_ctl value is irrelevant *)
        Delayed.return
          { value; observer = true; controller = false; obs_ctl = obs_value }
    | None ->
        Delayed.return
          {
            value = Expr.LVar (LVar.alloc ());
            obs_ctl = obs_value;
            observer = true;
            controller = false;
          }
  in
  MemMap.add pcy_id new_block pcy_env

let cons_controller pcy_env pcy_var =
  match MemMap.find_opt pcy_var pcy_env with
  | Some ({ controller = true; obs_ctl; _ } as pcy) ->
      (* We consume the pcy *)
      let new_pcy_env =
        MemMap.add pcy_var { pcy with controller = false } pcy_env
      in
      DR.ok (obs_ctl, new_pcy_env)
  | Some { controller = false; _ } -> DR.error (Err.Missing_controller pcy_var)
  | None -> DR.error (Err.Missing_pcy pcy_var)

let prod_controller pcy_env pcy_id ctl_value =
  let+ new_block =
    match MemMap.find_opt pcy_id pcy_env with
    | Some { controller = true; _ } ->
        Delayed.vanish () (* Duplicated resource *)
    | Some { value; obs_ctl; observer = true; controller = false } ->
        (* We already have an observer, we learn equality *)
        let learned = [ Formula.Infix.( #== ) ctl_value obs_ctl ] in
        Delayed.return ~learned
          { value; observer = true; controller = true; obs_ctl }
    | Some { value; observer = false; controller = false; obs_ctl = _ } ->
        (* We have neither the controller nor the observer, the obs_ctl value is irrelevant *)
        Delayed.return
          { value; observer = false; controller = true; obs_ctl = ctl_value }
    | None ->
        Delayed.return
          {
            value = Expr.LVar (LVar.alloc ());
            obs_ctl = ctl_value;
            observer = false;
            controller = true;
          }
  in
  MemMap.add pcy_id new_block pcy_env

let cons_value pcy_env pcy_id =
  (* The value is persistent information, it doesn't get consumed *)
  match MemMap.find_opt pcy_id pcy_env with
  | Some { value; _ } -> (value, pcy_env)
  | None ->
      (* If not there, we simply create it, and now we know its value. *)
      let value = Expr.LVar (LVar.alloc ()) in
      ( value,
        MemMap.add pcy_id
          {
            value;
            observer = false;
            controller = false;
            obs_ctl = Expr.Lit Null;
            (* The obs_ctl we assign is irrelevant, since we have neither observer nor controller *)
          }
          pcy_env )

let prod_value pcy_env pcy_id new_value =
  match MemMap.find_opt pcy_id pcy_env with
  | Some pcy ->
      (* `value` is a function, so if we already know its value, we now that whatever we have is equal *)
      Delayed.return
        ~learned:[ Formula.Infix.( #== ) pcy.value new_value ]
        pcy_env
  | None ->
      (* The obs_ctl we assign is irrelevant, since we have neither observer nor controller *)
      let block =
        {
          value = new_value;
          observer = false;
          controller = false;
          obs_ctl = Expr.Lit Null;
        }
      in
      Delayed.return (MemMap.add pcy_id block pcy_env)

let assign pcy_env pcy_var assigned =
  (* In order to assign an obs_ctl value, we need *both* the controller and observer at the same time. *)
  match MemMap.find_opt pcy_var pcy_env with
  | None -> DR.error (Err.Missing_pcy pcy_var)
  | Some { controller = false; _ } -> DR.error (Err.Missing_controller pcy_var)
  | Some { observer = false; _ } -> DR.error (Err.Missing_observer pcy_var)
  | Some ({ controller = true; observer = true; _ } as pcy) ->
      let new_pcies =
        MemMap.add pcy_var { pcy with obs_ctl = assigned } pcy_env
      in
      DR.ok new_pcies

let alloc pcy_env =
  let pcy_id = ALoc.alloc () in
  let value = Expr.LVar (LVar.alloc ()) in
  let obs_ctl = Expr.LVar (LVar.alloc ()) in
  let block = { value; obs_ctl; observer = true; controller = true } in
  let updated_env = MemMap.add pcy_id block pcy_env in
  (Expr.ALoc pcy_id, updated_env)

let pp ft t =
  let open Fmt in
  pf ft "@[<v 2>Prophecies:@,%a@]"
    (iter_bindings MemMap.iter (pair ~sep:(any " -> ") string pp_prophecy))
    t

let merge old_proph new_proph =
  let open Formula.Infix in
  let value_eq = old_proph.value #== new_proph.value in
  let controller = new_proph.controller || old_proph.controller in
  let observer = new_proph.observer || old_proph.observer in
  let obs_ctl, ctl_obs_eq =
    match
      ( old_proph.controller || old_proph.observer,
        new_proph.controller || new_proph.observer )
    with
    | true, true ->
        (* Both obs_ctl are relevant *)
        let ctl_obs_eq = [ old_proph.obs_ctl #== new_proph.obs_ctl ] in
        (new_proph.obs_ctl, ctl_obs_eq)
    | true, false ->
        (* Only old obs_ctl is relevant *)
        (old_proph.obs_ctl, [])
    | false, true ->
        (* Only new obs_ctl is relevant *)
        (new_proph.obs_ctl, [])
    | false, false ->
        (* Both obs_ctl are irrelevant *)
        (Expr.Lit Null, [])
  in
  let learned = value_eq :: ctl_obs_eq in
  DR.ok ~learned { value = new_proph.value; obs_ctl; observer; controller }

let substitution pcy_env subst =
  let open Gillian.Symbolic in
  if Subst.is_empty subst then DR.ok pcy_env
  else
    let loc_subst =
      Subst.fold subst
        (fun l r acc ->
          match l with
          | ALoc loc | Lit (Loc loc) -> (loc, r) :: acc
          | _ -> acc)
        []
    in
    let subst_expr = Subst.subst_in_expr subst ~partial:true in
    let values_substed =
      MemMap.map
        (fun block ->
          let value = subst_expr block.value in
          let obs_ctl = subst_expr block.obs_ctl in
          let controller = block.controller in
          let observer = block.observer in
          { value; obs_ctl; controller; observer })
        pcy_env
    in
    List.fold_left
      (fun acc (old_loc, new_loc) ->
        let** acc = acc in
        let new_loc =
          match new_loc with
          | Expr.Lit (Loc loc) | ALoc loc -> loc
          | _ ->
              Fmt.failwith
                "substitution failed, for location, target isn't a location"
        in
        match (MemMap.find_opt old_loc acc, MemMap.find_opt new_loc acc) with
        | None, None | None, Some _ -> DR.ok acc
        | Some prophecy, None ->
            MemMap.remove old_loc acc |> MemMap.add new_loc prophecy |> DR.ok
        | Some old_prophecy, Some new_prophecy ->
            let++ merged = merge old_prophecy new_prophecy in
            MemMap.remove old_loc acc |> MemMap.add new_loc merged)
      (DR.ok values_substed) loc_subst
