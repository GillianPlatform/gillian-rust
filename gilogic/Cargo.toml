[package]
name = "gilogic"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
gilogic_proc = { path = "../gilogic_proc" }
gilogic_proc_dummy = { path = "../gilogic_proc_dummy" }