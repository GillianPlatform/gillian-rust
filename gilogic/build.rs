fn main() {
    println!("cargo::rustc-check-cfg=cfg(gillian)");
}
