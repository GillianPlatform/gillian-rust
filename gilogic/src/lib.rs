#![allow(internal_features)]
#![feature(rustc_attrs)]
#![feature(never_type)]
#![feature(register_tool)]
#![feature(ptr_internals)]
#![feature(allocator_api)]
#![feature(tuple_trait, unboxed_closures)]
#![feature(stmt_expr_attributes, proc_macro_hygiene)]
#![register_tool(gillian)]

mod tys;

pub use tys::RustAssertion;

pub mod macros {
    pub use gilogic_proc::{
        assertion, assertion_test, borrow, close_borrow, extract_lemma, lemma, open_borrow,
        predicate, show_safety, specification,
    };
    pub mod prophecies {
        pub use gilogic_proc::with_freeze_lemma_for_mutref_pcy as with_freeze_lemma_for_mutref;
    }
    pub mod no_prophecies {
        pub use gilogic_proc::with_freeze_lemma_for_mutref_no_pcy as with_freeze_lemma_for_mutref;
    }
}

mod seq;
pub use seq::Seq;

mod ownable;
pub use ownable::Ownable;
pub mod alloc;
pub mod iterated;
pub mod prophecies;

pub mod symex_ctrl {

    #[gillian::builtin]
    #[gillian::no_translate]
    #[rustc_diagnostic_item = "gillian::symex_ctrl::symex_next"]
    #[inline(always)]
    pub fn symex_next() {}
}

#[path = "stubs.rs"]
pub mod __stubs;
