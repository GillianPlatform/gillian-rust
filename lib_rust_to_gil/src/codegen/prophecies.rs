use crate::prelude::*;

use rustc_middle::{mir::Place, ty::Ty};

use gillian::gil::Expr;

use super::context::GilCtxt;

mod action_names {
    pub(crate) const PCY_ALLOC: &str = "pcy_alloc";
}

pub enum ProphecyAction<'tcx> {
    Alloc(Ty<'tcx>),
}

impl<'tcx, 'body> GilCtxt<'tcx, 'body> {
    fn push_pcy_action(&mut self, target: String, action: ProphecyAction<'tcx>) {
        match action {
            ProphecyAction::Alloc(ty) => {
                let ty = self.encode_type(ty).into();
                self.push_cmd(Cmd::Action {
                    variable: target,
                    action_name: action_names::PCY_ALLOC.to_string(),
                    parameters: vec![ty],
                })
            }
        }
    }

    fn push_alloc_prophecy(&mut self, typ: Ty<'tcx>) -> Expr {
        let action = ProphecyAction::Alloc(typ);
        let target = self.temp_var();
        self.push_pcy_action(target.clone(), action);
        Expr::PVar(target)
    }

    pub fn push_create_prophecy_for(&mut self, place: Place<'tcx>) -> Expr {
        if !self.prophecies_enabled() {
            fatal!(self, "Prophecies are not enabled, something is wrong");
        }
        let ty = self.place_ty(place).ty;
        match self.global_env().get_repr_ty_for(ty) {
            Some(repr_ty) => self.push_alloc_prophecy(repr_ty),
            None => Expr::null(),
        }
    }
}
