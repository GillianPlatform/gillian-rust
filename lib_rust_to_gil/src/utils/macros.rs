macro_rules! fatal {
  ($e: expr, $($tts:tt)*) => {
      $e.ty_ctxt.sess.fatal(&format!($($tts)*))
  };
}

pub(crate) use fatal;
