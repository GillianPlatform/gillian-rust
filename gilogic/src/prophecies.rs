use crate as gilogic;
use crate::tys::RustAssertion;
use gilogic::macros::{assertion, borrow, lemma, predicate, specification};

macro_rules! unreachable {
    ($x:expr) => {
        panic!()
    };
    () => {
        panic!()
    };
}

/*
Here's an idea:
I could have a trait MutBorrow, which can be opened, closed etc...
And there would be functions between MutBorrows:
-> Splits which can split borrows in bits
-> pull, which can pull existentials from a borrow.

Not sure how that would work, but it'd certainly be a nicer interface.
*/

#[allow(non_snake_case)]
pub trait Ownable {
    #[rustc_diagnostic_item = "gillian::pcy::ownable::representation_ty"]
    type RepresentationTy;

    #[rustc_diagnostic_item = "gillian::pcy::ownable::own"]
    #[gillian::decl::pred_ins = "0"]
    fn own(self, model: Self::RepresentationTy) -> RustAssertion;

    // #[rustc_diagnostic_item = "gillian::pcy::ownable::ref_mut_inner"]
    // #[gillian::decl::pred_ins = "0"]
    // fn ref_mut_inner(&mut self) -> RustAssertion {
    //     unreachable!("Implemented in GIL")
    // }

    // #[rustc_diagnostic_item = "gillian::unfold::pcy::ownable::ref_mut_inner"]
    // fn ref_mut_inner_____unfold(&mut self) -> RustAssertion {
    //     unreachable!("Implemented in GIL")
    // }

    // #[rustc_diagnostic_item = "gillian::fold::pcy::ownable::ref_mut_inner"]
    // fn ref_mut_inner_____fold(&mut self) -> RustAssertion {
    //     unreachable!("Implemented in GIL")
    // }
}

macro_rules! own_int {
    ($t:ty) => {
        impl Ownable for $t {
            type RepresentationTy = $t;

            #[cfg(gillian)]
            #[predicate]
            fn own(self, model: $t) {
                assertion!(($t::MIN <= self) * (self <= $t::MAX) * (self == model))
            }

            #[cfg(not(gillian))]
            fn own(self, _model: $t) -> RustAssertion {
                unreachable!("")
            }
        }
    };

    ($t:ty, $($ts:ty),+) => {
        own_int!($t);
        own_int!($($ts),+);
    };
}

#[borrow]
fn mut_ref_inner_proph<'a, U: Ownable>(this: In<&'a mut U>) -> RustAssertion {
    assertion!(|v, repr| (this -> v) * v.own(repr) * controller(this.prophecy(), repr))
}

impl<T: Ownable> Ownable for &mut T {
    type RepresentationTy = (T::RepresentationTy, T::RepresentationTy);
    #[cfg(gillian)]
    #[predicate]
    fn own(self, model: (T::RepresentationTy, T::RepresentationTy)) {
        assertion!(|a, b| (model == (a, b))
            * mut_ref_inner_proph(self)
            * observer(self.prophecy(), a)
            * (self.prophecy().value() == b))
    }

    #[cfg(not(gillian))]
    fn own(self, _model: (T::RepresentationTy, T::RepresentationTy)) -> RustAssertion {
        unreachable!("")
    }
}

own_int!(u8, u16, u32, u64, u128, usize, i8, i16, i32, i64, i128, isize);

impl<T: Ownable, U: Ownable> Ownable for (T, U) {
    type RepresentationTy = (T::RepresentationTy, U::RepresentationTy);

    #[cfg(gillian)]
    #[predicate]
    fn own(self, model: (T::RepresentationTy, U::RepresentationTy)) {
        assertion!(|l, r| (model == (l, r)) * self.0.own(l) * self.1.own(r))
    }

    #[cfg(not(gillian))]
    fn own(self, _model: (T::RepresentationTy, U::RepresentationTy)) -> RustAssertion {
        unreachable!("Implemented in GIL");
    }
}

impl<T: Ownable> Ownable for Option<T> {
    type RepresentationTy = Option<T::RepresentationTy>;

    #[cfg(gillian)]
    #[predicate]
    fn own(self, model: Option<T::RepresentationTy>) {
        assertion!((self == None) * (model == None));
        assertion!(|ax: T, repr: _| (self == Some(ax))
            * <T as Ownable>::own(ax, repr)
            * (model == Some(repr)))
    }

    #[cfg(not(gillian))]
    fn own(self, _model: Option<T::RepresentationTy>) -> RustAssertion {
        unreachable!("Implemented in GIL");
    }
}

impl Ownable for () {
    type RepresentationTy = ();

    #[cfg(gillian)]
    #[predicate]
    fn own(self, model: ()) -> RustAssertion {
        assertion!((self == ()) * (self == model))
    }

    #[cfg(not(gillian))]
    fn own(self, _model: ()) -> RustAssertion {
        unreachable!("Implemented in GIL");
    }
}

pub trait Prophecised<P> {
    #[gillian::no_translate]
    #[gillian::builtin]
    #[rustc_diagnostic_item = "gillian::mut_ref::get_prophecy"]
    fn prophecy(self) -> Prophecy<P>;

    #[gillian::no_translate]
    #[gillian::builtin]
    #[rustc_diagnostic_item = "gillian::mut_ref::set_prophecy"]
    fn with_prophecy(self, pcy: Prophecy<P>) -> Self;
}

impl<T> Prophecised<T::RepresentationTy> for &mut T
where
    T: Ownable,
{
    fn prophecy(self) -> Prophecy<T::RepresentationTy> {
        unreachable!("Implemented in GIL")
    }

    fn with_prophecy(self, _pcy: Prophecy<T::RepresentationTy>) -> Self {
        unreachable!("Implemented in GIL")
    }
}

#[specification(forall v, a, r.
    requires {
        (p -> v) * v.own(r) *
        observer(p.prophecy(), a) *
        controller(p.prophecy(), a)
    }
    ensures {
        (p -> v) * v.own(r) *
        observer(p.prophecy(), r) *
        controller(p.prophecy(), r)
}
)]
#[gillian::timeless]
pub fn prophecy_auto_update<T: Ownable>(p: &mut T) {
    let _ = p;
    unreachable!();
}

#[specification(forall m.
    requires { p.own(m) }
    ensures { $(m.0 == m.1)$ }
)]
#[gillian::timeless]
pub fn prophecy_resolve<T: Ownable>(p: &mut T) {
    let _ = p;
    unreachable!();
}

#[gillian::builtin]
#[rustc_diagnostic_item = "gillian::prophecy::check_obs_sat"]
pub fn check_obs_sat() {
    unreachable!();
}

#[derive(Clone, Copy)]
pub struct Prophecy<T: ?Sized>(core::marker::PhantomData<T>);

impl<T> Prophecy<T> {
    #[gillian::no_translate]
    #[gillian::builtin]
    #[rustc_diagnostic_item = "gillian::prophecy::get_value"]
    pub fn value(self) -> T {
        unreachable!("Implemented in GIL")
    }

    #[gillian::no_translate]
    #[gillian::builtin]
    #[rustc_diagnostic_item = "gillian::prophecy::resolve"]
    pub fn resolve(self) {
        unreachable!("Implemented in GIL")
    }

    #[gillian::no_translate]
    #[gillian::builtin]
    #[rustc_diagnostic_item = "gillian::prophecy::assign"]
    pub fn assign(self, _v: T) {
        unreachable!("Implemented in GIL")
    }

    #[gillian::no_translate]
    #[gillian::builtin]
    #[rustc_diagnostic_item = "gillian::prophecy::assign_proph"]
    pub fn assign_proph(self, _v: Prophecy<T>) {
        unreachable!("Implemented in GIL")
    }
}

#[gillian::no_translate]
#[gillian::builtin]
#[rustc_diagnostic_item = "gillian::prophecy::controller"]
pub fn controller<T>(_x: Prophecy<T>, _v: T) -> RustAssertion {
    unreachable!("")
}

#[gillian::no_translate]
#[gillian::builtin]
#[rustc_diagnostic_item = "gillian::prophecy::observer"]
pub fn observer<T>(_x: Prophecy<T>, _v: T) -> RustAssertion {
    unreachable!("")
}

#[macro_export]
macro_rules! mutref_auto_resolve {
    ($x: expr) => {
        gilogic::prophecies::check_obs_sat();
        gilogic::prophecies::prophecy_auto_update($x);
        gilogic::prophecies::prophecy_resolve($x);
    };
}
