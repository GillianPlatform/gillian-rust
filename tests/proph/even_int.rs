//@check-pass
extern crate creusillian;
extern crate gilogic;

use gilogic::macros::{assertion, predicate};
use gilogic::mutref_auto_resolve;
use gilogic::prophecies::{Ownable, Prophecised};

struct EvenInt {
    num: i32,
}

impl Ownable for EvenInt {
    type RepresentationTy = i32;

    #[predicate]
    fn own(self, model: i32) {
        assertion!((self == EvenInt { num: model }) * (model % 2 == 0));
    }
}

impl EvenInt {
    #[creusillian::ensures(true)]
    pub fn new_2(x: i32) -> Self {
        if x % 2 == 0 {
            Self { num: x }
        } else {
            if x < 1000 {
                Self { num: x + 1 }
            } else {
                Self { num: x - 1 }
            }
        }
    }

    pub unsafe fn new(x: i32) -> Self {
        Self { num: x }
    }

    #[creusillian::ensures(true)]
    pub fn new_3(x: i32) -> Option<Self> {
        if x % 2 == 0 {
            let y = unsafe { Self::new(x) };
            Some(y)
        } else {
            None
        }
    }

    unsafe fn add(&mut self) {
        self.num += 1;
    }

    #[creusillian::ensures(true)]
    pub fn test(&mut self) {
        if self.num % 2 != 0 {
            panic!()
        }
    }

    #[creusillian::requires((*self@) <= i32::MAX@ - 2)]
    #[creusillian::ensures((^self@) == (*self@) + 2)]
    pub fn add_two(&mut self) {
        self.num;

        unsafe {
            self.add();
            self.add();
        }
        mutref_auto_resolve!(self);
    }
}
