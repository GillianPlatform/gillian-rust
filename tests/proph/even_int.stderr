warning: unused import: `Prophecised`
 --> ../tests/proph/even_int.rs:7:36
  |
7 | use gilogic::prophecies::{Ownable, Prophecised};
  |                                    ^^^^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default

[TIMESTAMP WARN  lib_rtg::metadata] could not read metadata for crate `"std"`: Os { code: 2, kind: NotFound, message: "No such file or directory" }
[TIMESTAMP WARN  lib_rtg::metadata] could not read metadata for crate `"creusillian"`: Os { code: 2, kind: NotFound, message: "No such file or directory" }
warning: struct `EvenInt` is never constructed
 --> ../tests/proph/even_int.rs:9:8
  |
9 | struct EvenInt {
  |        ^^^^^^^
  |
  = note: `#[warn(dead_code)]` on by default

warning: 2 warnings emitted

[TIMESTAMP INFO  rust_to_gil] Compiled!
